lib/src/cancelable_operation.dart

```diff
@@ -150,6 +150,65 @@ class CancelableOperation<T> {
       {FutureOr<R> Function(Object, StackTrace)? onError,
       FutureOr<R> Function()? onCancel,
       bool propagateCancel = false}) {
+    if (onError != null && onCancel != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onError: (error, stack, completer) {
+        completer.complete(onError(error, stack));
+      }, onCancel: (completer) {
+        completer.complete(onCancel());
+      }, propagateCancel: propagateCancel);
+    } else if (onError != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onError: (error, stack, completer) {
+        completer.complete(onError(error, stack));
+      }, propagateCancel: propagateCancel);
+    } else if (onCancel != null) {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, onCancel: (completer) {
+        completer.complete(onCancel());
+      }, propagateCancel: propagateCancel);
+    } else {
+      return thenOperation<R>((value, completer) {
+        completer.complete(onValue(value));
+      }, propagateCancel: propagateCancel);
+    }
+  }
+
+  /// Creates a new cancelable operation to be completed
+  /// when this operation completes or is cancelled.
+  ///
+  /// The [onValue] and [onError] callbacks behave in the same way as
+  /// for [Future.then], and the result of those callbacks is used to complete
+  /// the returned cancelable operation.
+  ///
+  /// If this operation completes normally [onValue] is called with the result
+  /// and the [CancelableCompleter] controlling the returned operation.
+  ///
+  /// If this operation completes as an error, and an [onError] callback is
+  /// provided, it is called with the error, stack trace, and the
+  /// [CancelableCompleter] controlling the returned operation.
+  ///
+  /// If this operation is canceled, and an [onCancel] call is provided, it is
+  /// called with the [CancelableCompleter] controlling the returned operation.
+  ///
+  /// If [onCancel] is not provided, and this operation is canceled, then the
+  /// returned operation is also canceled.
+  ///
+  /// At most one of [onValue], [onError], or [onCancel] will be called.
+  /// If any of [onValue], [onError], or [onCancel] cause any asynchronous
+  /// errors, including through a returned `Future`, they will not be handled or
+  /// forwarded through the returned operation.
+  ///
+  /// If [propagateCancel] is `true` and the returned operation is canceled
+  /// then this operation is canceled. The default is `false`.
```

Comment:
@lrhn  I tried to go back and understand the motivation for this based on https://github.com/dart-lang/async/pull/81#issuecomment-479444606 but I'm not sure I understand.

I can't think of the reason you _wouldn't_ want to propagate cancellation back to `this` from the returned operation. I was tempted to not have this argument, or change the default, because in each usage I've encountered so far it's felt like extra noise. Would it be better to keep consistency even if most use cases need to pass the argument?

Do you recall what use case we expected where you wouldn't want the cancellation to propagate?

Reply:
No recollection. My guess is that it was added later, and therefore the default was made the previous behavior.

We can do a code search to see if anyone ever uses the default. If everybody passes `true` anyway, we might get away with changing the default (in a major version increment, obviously).

Reply:
>  My guess is that it was added later, and therefore the default was made the previous behavior.

That was my original assumption, but it turns out the parameter and default were there from the start. The only discussion I can find about propagation is the thread I linked above (Also mentioned [here](https://github.com/dart-lang/async/pull/83#issuecomment-480074557) but without any new context). It has a bit of discussion of both forwards propagation (`this` gets cancelled so  the returned operation gets canceled) and backwards propagation (the `propagateCancel` behavior), but I don't understand how exactly the discussion lead to the code we shipped.

> We can do a code search to see if anyone ever uses the default.

There are zero usages I can find of the argument being passed, either internally or on github. (Maybe no one really cares?)

There are many uses where I don't think it matters.
Two common patterns are:

- `return CancelableOperation.fromFuture(someFuture).then(transformResult)` - the cancelation propagation doesn't have an observable impact. Only the transform work can be canceled, `someFuture` happens either way.
- `someOperation().then(useResult);` - the returned operation is dropped and can't be cancelled, so propagation never matters. Could have been written as `someOperation().value.then(useResult)` but the former is shorter.

I also see uses where I suspect the cancellation propagation is what was intended, but was maybe never tested. `return someOperation().then(transformResult);`.  Since `someOperation()` returns a `CancelableOperation` I'd expect the intention is for the caller to be able to cancel it - and this caller will never cancel it (but the author might think they are).

I have not yet seen a case that strikes me as one where an operation could have multiple uses that need to avoid canceling each other.

I can try changing the default and see if there happen to be any tests for relying on this behavior.